lighthouse-core/audits/redirects.js

```diff
@@ -20,68 +20,92 @@ class Redirects extends Audit {
-          });
-        }
-
-        const headings = [
-          {key: 'url', itemType: 'text', text: 'Redirected URL'},
-          {key: 'wastedMs', itemType: 'ms', text: 'Time for Redirect', granularity: 1},
-        ];
-        const summary = {wastedMs: totalWastedMs};
-        const details = Audit.makeTableDetails(headings, pageRedirects, summary);
-
-        return {
-          // We award a passing grade if you only have 1 redirect
-          score: redirectRequests.length <= 2 ? 1 : UnusedBytes.scoreForWastedMs(totalWastedMs),
-          rawValue: totalWastedMs,
-          displayValue: Util.formatMilliseconds(totalWastedMs, 1),
-          extendedInfo: {
-            value: {
-              wastedMs: totalWastedMs,
-            },
-          },
-          details,
-        };
+  static async audit(artifacts, context) {
+    const trace = artifacts.traces[Audit.DEFAULT_PASS];
+    const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+    const simulatorData = {devtoolsLog, settings: context.settings};
+    const simulator = await artifacts.requestLoadSimulator(simulatorData);
+
+    // TODO(phulce): split MetricComputationInput type and avoid this cast
+    const metricSettings = /** @type {LH.Config.Settings} */ ({throttlingMethod: 'simulate'});
+    const metricComputationData = {trace, devtoolsLog, simulator, settings: metricSettings};
+
+    const mainResource = await artifacts.requestMainResource({URL: artifacts.URL, devtoolsLog});
+    const metricResult = /** @type {LH.Artifacts.LanternMetric} */ (
+      await artifacts.requestInteractive(metricComputationData));
+
+    /** @type {Map<string, LH.Gatherer.Simulation.NodeTiming>} */
+    const nodeTimingsByUrl = new Map();
+    for (const [node, timing] of metricResult.pessimisticEstimate.nodeTimings.entries()) {
+      if (node.type === 'network') {
+        const networkNode = /** @type {LH.Gatherer.Simulation.GraphNetworkNode} */ (node);
+        nodeTimingsByUrl.set(networkNode.record.url, timing);
+      }
+    }
+
+    // redirects is only available when redirects happens
+    const redirectRequests = Array.from(mainResource.redirects || []);
+
+    // add main resource to redirectRequests so we can use it to calculate wastedMs
+    redirectRequests.push(mainResource);
+
+    let totalWastedMs = 0;
+    const pageRedirects = [];
+
+    // Kickoff the results table (with the initial request) if there are > 1 redirects
+    if (redirectRequests.length > 1) {
+      pageRedirects.push({
+        url: `(Initial: ${redirectRequests[0].url})`,
+        wastedMs: 0,
+      });
+    }
+
+    for (let i = 1; i < redirectRequests.length; i++) {
+      const initialRequest = redirectRequests[i - 1];
+      const redirectedRequest = redirectRequests[i];
+
+      const initialTiming = nodeTimingsByUrl.get(initialRequest.url);
+      const redirectedTiming = nodeTimingsByUrl.get(redirectedRequest.url);
+      if (!initialTiming || !redirectedTiming) continue;
```

Comment:
wouldn't this be a error with the graph if any of the redirects aren't in there? Maybe we should throw or log an error in that (unlikely?) case?

Reply:
fair, I'll switch to error